### PR TITLE
Create nce_cost/README_en.md

### LINKED ISSUE #719: Translation Plan-使用噪声对比估计加速语言模型训练-汉译英
## 任务背景
针对v2版本，目前我们已有很多中文使用文档，但仍有大量的汉译英工作及少量英译汉工作没有完成，我们希望将中英文版本补充完全，提升用户的阅读体验，也有益于国外用户在使用时查阅文档。
此外，我们的模型库Models，仍有大部分英译汉和汉译英的工作，亟待完成。

## 任务描述
1. 需要做什么：
把“使用噪声对比估计加速语言模型训练”**汉译英**

2. 地址
- 中文地址： [Github地址](https://github.com/PaddlePaddle/models/blob/develop/nce_cost/README.md)，[网页地址](http://www.paddlepaddle.org/docs/develop/models/nce_cost/README.html)
- 需要翻译的英文地址，暂无，可在中文版所在github目录下创建md文档

## 如何认领
我们欢迎大家积极认领，内部小伙伴请通过self assign的方式认领，外部小伙伴可通过在本issue下方comment认领。

## 注意事项
上交初稿的截止日期是2018年4月15日
上交后，PaddlePaddle的RD会为大家review
请于review后的5个工作日之内完成修改

## 相关信息
所有需要翻译的文档：
V2使用文档：
https://shimo.im/sheet/PqK407VYm80Ju3dw
Models:
https://shimo.im/sheet/6XTZnRQEUS0uC1Ut

### PR DESCRIPTION
**Pull request by shadowimp**
Update nce_cost/README.md
fix #719

### CODE REVIEW

**Comment by shanyi15** [comment]
看起来还是Trim Trailing Whitespace挂了，请问您在sublime里是否有设置
"trim_trailing_white_space_on_save": true, 
"ensure_newline_at_eof_on_save": true,
这两句呢？

**Review by shadowimp — COMMENTED** [review]
fix Trim trailing whitespace using sublime the second time

**Comment by shanyi15** [comment]
@shadowimp 您好，发现一个很严重的问题。您直接在中文版文档的基础上改了，而不是提交了一个新文件，这样一旦merge了以后，中文版就被英文版覆盖了，请您以提交新文件的方式提交这篇翻译文档，感谢！

**Comment by shanyi15** [comment]
hi，可能之前我没表述清楚，抱歉。
现在已有的readme.md文档是中文版的，我们需要做的是，翻译成英文并命名为readme_en.md上传到相应的目录下
请注意新文件的名字是README_en.md

**Review by shadowimp — COMMENTED** [review]
add README.cn

**Comment by shanyi15** [comment]
请不要改动原中文版文件的名字哈，中文版保持README.md
英文版命名为README_en.md
谢谢！

**Review by shadowimp — COMMENTED** [review]
create README.en

**Review by shadowimp — COMMENTED** [review]
update README_en.md

**Comment by shanyi15** [comment]
你好，文档中的配图是中文的，请翻译成英文，请将新图片上传到models/nce_cost/images/，原图片保持不动，新图片命名为network_conf_en.png，并在文档中引这个英文版的图片。
谢谢！

**Review by shadowimp — COMMENTED** [review]
update nce_cose : add network_conf_en.png

File: nce_cost/README_en.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+| param\_attr / bias\_attr | use to set parameter name  |  facilitates the loading of parameters during the forecast phase, as described in the prediction section                                                                                                   |
+| num\_neg\_samples        |  Number of Negative Samples                  |  You can control the proportion of positive and negative samples. The range of this value is [1, dictionary size -1]. The more negative samples, the slower the training of the whole model. The accuracy of the model will also be higher                                  |
+| neg\_distribution        |  Generates the distribution of negative sample tags. The default is a uniform distribution| You can control the sampling weights for each category when negative samples are sampled. For example, if you want the positive example to be “sunny,” and the negative example “flood” is highlighted more during training, you can increase the weight of the “flood” sample |
+
```
Comment (skylarch) [inline nce_cost/README_en.md:115]: # Accelerating the Training Process of a Language Model with Noise Contrast Estimation

## Why We Need Noise Comparison

A language model is the basis of many natural language processing tasks, and it is also an effective method for obtaining word vector representations. The Neural Probabilistic Language Model (NPLM) characterizes the probability of a sequence of words $\omega_1,...,\omega_T$ belonging to a fixed language $P(\omega_1^T)$ :
$$P(\omega_1^T)= \prod_{t=1}^{T}P(\omega_t|\omega_1^{t-1})$$

In order to reduce the difficulty of modeling and solving, we often introduce certain conditional independent assumptions: the probability of the word $w_t$ is only affected by the previous $n-1$ words, so there is formula ($1$):

$$ P(\omega_1^T) \approx \prod P(\omega_t|\omega_{t-n-1}^{t-1}) \tag{1}$$

From the formula ($1$) we can see that the entire sequence $\omega_1,...,\omega_T$ can be calculated by the modeling conditional probabilities $P(\omega_t|w_{t-n-1},...,\omega_{t-1})$. So, we can summarize the task of language model solving as follows:

** Giver the vector representation or the context, $h$, of a word sequence, let the model predict the probability of the next target word $\omega$. **

In the [$n$-gram language model] (https://github.com/PaddlePaddle/book/tree/develop/04.word2vec), the context takes a fixed $n-1$ words. [RNN language model] (https://github.com/PaddlePaddle/models/tree/develop/generate_sequence_by_rnn_lm) can handle any length of context.

Given context $h$, NPLM learns a scoring function $s_\theta(\omega, h)$, which characterizes the similarities between the vector of $h$ and \omega'$, the vector of the possible words coming next. The similarity between \omega'$ is then normalized by dividing the value of the scoring function $s$ in the whole word table space (divided by the normalization factor $Z$) to get the probability distribution of the target word $\omega$. $\theta$ is a learning parameter, this process is expressed by formula ($2$), the calculation of `Softmax`.

$$P_\theta^h(\omega) = \frac{\text{exp}{s_\theta(\omega, h)}}{Z},Z=\sum_{\omega'} \exp{s_\ Theta(\omega', h)}\tag{2}$$

MLE (Maximum Likelihood Estimation) is the most commonly used learning criterion for solving probability ($2$). However, the normalization factor $Z$ must be calculated whether the estimated probability $P_\theta^h(\omega)$ or the gradient of the likelihood is calculated. The calculation of $Z$ increases linearly with the size of the dictionary. When training large-scale language models, such as a dictionary with a size of a million or more, the training time will become very large. Therefore, we need other possible learning guidelines with solution process easier and more solvable. **

Another article of models intruduces the use of [Hsigmoid Accelerated Word Vector Training] (https://github.com/PaddlePaddle/models/tree/develop/hsigmoid). Here, we introduce sample-based training model, Noise-contrastive estimation (NCE)\[[1](#ref)\], to speed up language training.

## What Noise Comparison Estimation Is

The noise comparison estimation is a probability density estimation criterion based on the sampling idea and is used for estimation/fitting: the probability function is composed of a non-normalized score function and a normalization factor[[1](#References)\]. The noise comparison estimation avoids the calculation of the normalization factor $Z$ in the full dictionary space by constructing an auxiliary problem such as the following, which reduces the computational cost:

Given context $h$ and any known noise distribution $P_n$, learn a two-class classifier to fit: target $\mega$ from true distribution $P_\theta$ ($D = 1$) or the probability of noise distribution, $P_n$ ($D = 0$). Assuming that the number of negative samples from the noise distribution is $k$ times the target sample, then there are:

$$P(D=1|h,\omega) = \frac{P_\theta(h, \omega)}{P_\theta (h, \omega) + kP_n} \tag{3}$$

We directly use the `Sigmoid` function to characterize a binary classification probability ($3$):

$$P(D=1|h,\omega) = \sigma (\Delta s_\theta(w,h)) \tag{4}$$

With the above problem set, maximum likelihood estimation can be performed based on two classifications: increasing the probability of positive samples and reducing the probability of negative samples [[2,3] (#references)], that is, minimizing the following loss function:

$$
J^h(\theta )=E_{ P_d^h }\left[ \log {P^h(D=1|w,\theta )} \right] +kE_{ P_n }\left[ \log P^h (D=0|w,\theta )\right]$$
$$
 \\\\\qquad =E_{ P_d^h }\left[ \log { \sigma (\Delta s_\theta(w,h)) } \right] +kE_{ P_n }\left[ \log (1- \sigma (\Delta s_\theta(w,h))) \right] \tag{5}$$

The formula ($5$) is the NCE loss function defined based on noise comparison estimation. At this point, we have two remaining questions:
1. What is $s_\theta(w,h)$ in the formula ($5$)?
    - In a neural network implementation, $s_\theta(h,\omega)$ is an unnormalized score.
    - The learnable parameter $W$ of the NCE cost layer is a matrix of $|V| \times d$ dimensions; $|V|$ is the dictionary size, and $d$ is the dimension of the context vector $h$;
    - During training, the actual category of the next word, $t$, is a positive class. Sampling $k$ negative samples from the specified noise distribution are classified as follows: $\{n_1, ..., n_k\}$;
    - Extract the $\{t, n_1, ..., n_k\}$th line in $W$ ($k + 1$ lines in total) and calculate the points $s_\theta(w,h)$ with $h$ respectively. The final loss is calculated by ($5$).
2. How to choose the noise distribution?
    - In practice, the noise distribution can be arbitrarily chosen (the noise distribution implies a certain priori).
    - The most common choices are: use the unbiased uniform distribution, `unigram`, based on a full dictionary (word frequency statistics).
    - If the user does not specify a noise distribution in the PaddlePaddle, the default is to use a uniform distribution.

When using NCE to accurately train, the calculation cost of the last layer is only linearly related to the number of negative samples. When the number of negative samples gradually increases, the NCE estimation criterion converges to the maximum likelihood estimation. Therefore, when training using the NCE criterion, the quality of the normalized probability distribution can be controlled by changing the number of negative samples.

## Experimental Data
This example uses a Penn Treebank (PTB) data set ([Tomas Mikolov pre-processed version] (http://www.fit.vutbr.cz/~imikolov/rnnlm/simple-examples.tgz)) to train a 5-gram language model. PaddlePaddle provides a [paddle.dataset.imikolov](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/dataset/imikolov.py) interface to facilitate the use of PTB data. When no downloaded data is found, the script will automatically downloads and verifies the integrity of the file. The corpus language is in English, with a total of 42068 sentence training data and 3761 sentence test data.

## Network Structure
The detailed network structure of the 5-gram neural probabilistic language model is shown in Figure 1:

<p align="center">
<br/>
Figure 1. 5-gram network configuration structure
</p>

The model is mainly divided into the following parts:

1. **Input Layer**: the input sample consists of the original words, and each word is first converted to the ID in the dictionary.

2. **Word vector layer **: IDs are transformed to continously expressed word vector representation, which can better show the semantic relationship between words. After the training is completed, the semantic similarity between words can be expressed using the distance between the word vectors. The more similar the semantics, the closer the distance.

3. **Word vector splice layer**: concatenates the word vectors and connects the word vectors end to end to form a long vector. This can facilitate the processing of the fully connected layer.

4. **FULL CONNECTION Hidden Layer**: inputs the long vector obtained from the previous layer into a hidden layer of neural network and outputs the feature vector. A fully connected hidden layer can enhance the learning ability of the network.

5. **NCE layer**: The paddle.layer.nce` provided by PaddlePaddle can be used directly as a loss function during training.


## Training
Run ``` python train.py ``` in the command line window to start the training task.

- The first time the program runs, it will detect if the PTB data set is included in the user's cache folder. If it is not included, it will be automatically downloaded.
- Every 10 batch prints the value of model training on the training set during the run
- After each pass is completed, the loss on the test data set is calculated and the latest snapshot of the model is also saved.

The NCE code in the model file, `network_conf.py`, is as follows:

```python
Return paddle.layer.nce(
            Input=hidden_layer,
            Label=next_word,
            Num_classes=dict_size,
            Param_attr=paddle.attr.Param(name="nce_w"),
            Bias_attr=paddle.attr.Param(name="nce_b"),
            Num_neg_samples=25,
            Neg_distribution=None)
```

Some important parameters of the NCE layer are explained as follows:

Parameter Name | Parameter Function | Introduction |

File: nce_cost/README_en.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+     0.6734  their   may want to move
+     ```
+
+     Each row is a forecast result, internally separated by "\t", for a total of 3 columns:
+     - First column: Probability of the next word.
+     - Second column: The next word of the model prediction.
+     - Third column: $n$ words entered, separated by spaces.
+
+
+## Reference
```
Comment (skylarch) [inline nce_cost/README_en.md:156]: ## Prediction
1. Run from the command line window:
    ```bash
    Python infer.py \
      --model_path "models/XX" \
      --batch_size 1 \
      --use_gpu false \
      --trainer_count 1
    ```
    The parameters are as follows:
    - `model_path`: specifies the path where the trained model is located. This parameter is required.
    - `batch_size`: predicts the number of samples in parallel at a time. Optional, the default value is `1`.
    - `use_gpu`: determines whether or not to use GPU for prediction. Optional, the default value is `False`.
    - `trainer_count` : The number of threads to use for prediction. Optional, defaults to `1`. **Note: The number of threads used for prediction must be greater than the number of samples to be predicted concurrently**.

2. It should be noted that** the calculation logic for prediction and training is different**. During the prediction phase, the model uses full-joined matrix multiplication followed by `softmax` activation and outputs the probability distribution for each category. The model needs to replace the `paddle.train.nce` layer used in the training. In PaddlePaddle, the NCE layer stores the learnable parameters as a matrix of `[number of classes × output vector width of the previous layer]`. When predicting, **full-joined operations need to be transferred when loading the NCE layer to learn the parameters. Set **, the code is as follows:
    ```python
    Return paddle.layer.mixed(
          Size=dict_size,
          Input=paddle.layer.trans_full_matrix_projection(
              Hidden_layer, param_attr=paddle.attr.Param(name="nce_w")),
          Act=paddle.activation.Softmax(),
          Bias_attr=paddle.attr.Param(name="nce_b"))
    ```
     `paddle.layer.mixed` in the above code snippet must be entered as `paddle.layer.×_projection` in PaddlePaddle. `paddle.layer.mixed` sums up multiple `projection` (input can be multiple) calculations as output. `paddle.layer.trans_full_matrix_projection` transposes parameter $W$ when calculating matrix multiplication.

3. The predict output format is as follows:
    ```text
    0.6734 their may want to move
    ```

    Each row is a prediction result with a total of 3 columns separated by "\t":
    - First column: probability of the next word.
    - Second column: the next word of the model prediction.
    - Third column: $n$ words entered, separated by spaces.


## References

File: nce_cost/README_en.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
+Running the following program need use PaddlePaddle v0.10.0 version。If your PaddlePaddle installed lower demand，Please following the [install  document ](http://www.paddlepaddle.org/docs/develop/documentation/zh/build_and_install/pip_install_cn.html) instruction upgrade PaddlePaddle install version。
```
Comment (skylarch) [inline nce_cost/README_en.md:1]: Running sample code in this directory requires PaddelPaddle v0.10.0 and later. If the PaddlePaddle on your device is lower than this version, please follow the instructions in [installation document](http://www.paddlepaddle.org/docs/develop/documentation/zh/build_and_install/pip_install_cn.html) and make an update.

**Comment by skylarch** [comment]
根据最新的格式要求，请将英文README文档的名称改为README.md。
